utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/collection.scala

```diff
@@ -0,0 +1,135 @@
+  = ListMap(map.asScala.toSeq.sortBy(_._2)(implicitly[Ordering[V]].reverse):_*).asJava
+
+  @Documentation(description = "Returns a list made of first n elements of the given list")
+  @GenericType(typingFunction = classOf[TypingFunctions.ListTyping])
+  def take[T](@ParamName("list") list: java.util.List[T], @ParamName("max") max: Int): java.util.List[T]
+  = list.subList(0, Math.min(max, list.size()))
+
+  @Documentation(description = "Creates a string made of all elements of the list separated with the given separator")
+  def join[T](@ParamName("separator") separator: String,  @ParamName("list") list: java.util.List[T]): String
+  = String.join(separator, list.asScala.map(Objects.toString).asJava)
+
+  @Documentation(description = "Cross joins two lists of maps: eg. product({{a: 'a'},{b: 'b'}}, {{c: 'c'},{d: 'd'}}) => {{a: 'a',c: 'c'},{b: 'b',c: 'c'},{a: 'a',d: 'd'},{b: 'b',d: 'd'}}")
+  def product[K, V](list1: java.util.List[java.util.Map[K, V]], list2: java.util.List[java.util.Map[K, V]]): java.util.List[java.util.Map[K, V]] = {
+    val l1 = list1.asScala.map(_.asScala)
+    val l2 = list2.asScala.map(_.asScala)
+    val res = for {
+      m1 <- l1
+      m2 <- l2
+    } yield m1 ++ m2
+    res.map(_.asJava).asJava
+  }
+
+  @Documentation(description = "Returns a list that contains all elements contained in list1, that doesn't appear in list2")
+  @GenericType(typingFunction = classOf[TypingFunctions.ListTyping])
+  def diff[T](@ParamName("list1") list1: java.util.List[T], @ParamName("list2") list2: java.util.List[T]): java.util.List[T] = {
+    list1.asScala.filterNot(list2.asScala.toSet).asJava
+  }
+
+  @Documentation(description = "Returns a list that contains all unique elements that are contained by both list1 and list2")
+  @GenericType(typingFunction = classOf[TypingFunctions.ListTyping])
+  def intersect[T](@ParamName("list1") list1: java.util.List[T], @ParamName("list2") list2: java.util.List[T]): java.util.List[T] = {
+    (list1.asScala.toSet intersect list2.asScala.toSet).toList.asJava
+  }
+
+  @Documentation(description = "Returns a list that contains unique elements from the given list")
+  @GenericType(typingFunction = classOf[TypingFunctions.ListTyping])
+  def distinct(@ParamName("list") list: java.util.List[Any]): java.util.List[Any] = {
+    new java.util.ArrayList(new java.util.HashSet[Any](list))
+  }
+
+  @Documentation(description = "Returns a copy of the list with its elements shuffled")
+  @GenericType(typingFunction = classOf[TypingFunctions.ListTyping])
+  def shuffle[T](@ParamName("list") list: java.util.Collection[T]): java.util.List[T] = {
+    val values = new java.util.ArrayList[T](list)
+    Collections.shuffle(values)
+    values
+  }
+
+  // spel *does* auto conversion between object and collection (`a` -> {`a`}).
+  // Unfortunately, sometimes spel converts object to *set* what causes unexpected application distinct to collection.
```

Comment:
That's... not English :D

And I don't see this `org.p4.esp.process.services.FlattenCollectionHelperSpec` test.

Reply:
Seems like the whole comment no longer applies. I've changed the argument type to `java.util.Collection` and works as expected